mla/analysis.py

```diff
@@ -323,47 +323,522 @@ def produce_and_minimize(self, event_model: models.EventModel, # Super class wil
+                self.Spatial_SoB = self.Spatial_S/self.Spatial_B
+                self.drop = self.N - mask.sum()
+                self.first_spatial = False
+                if self.injector.signal_time_profile is None:
+                    Time_S = 1
+                    Time_B = 1
+                else:
+                    Time_S = self.injector.signal_time_profile.pdf(self.data['time'])
+                    Time_B = self.injector.background_time_profile.pdf(self.data['time'])
+                self.Time_SOB = Time_S/Time_B
+                self.Time_SOB[np.isnan(self.Time_SOB)] = 0
+                if np.isinf(self.Time_SOB).sum() > 0:
+                    print("Warning:Background time profile doesn't not cover signal time profile.Discard events outside background window!")
+                    time_SoB[np.isinf(self.time_SoB)] = 0
+            self.energy_SoB = event_model.get_energy_sob(self.data)
+        
+        def cal_ts(ns):
+            ts =( ns/self.N * (self.energy_SoB*self.Spatial_SoB*self.Time_SOB - 1))+1
+            ts_value = -2*(np.sum(np.log(ts))+self.drop*np.log(1-ns/self.N))
+            return ts_value
+        result = scipy.optimize.minimize(cal_ts,
+                    x0 = [1,],
+                    bounds = bounds,
+                    method = 'SLSQP',
+                    )     
+        fit_ns = result.x[0]
+        fit_ts = -1*result.fun
+        if np.isnan(fit_ts):
+            fit_ts = 0
+            fit_ns = 0
+        return fit_ts,fit_ns             
+        
 class TimeDependentPsAnalysis(Analysis):
     """Docstring"""
 
     def __init__(self,
-                 source: Optional[Dict[str, float]] = None, # Python 3.9 bug... pylint: disable=unsubscriptable-object
-                 bg_time_profile: Optional[time_profiles.GenericProfile] = None, # Python 3.9 bug... pylint: disable=unsubscriptable-object
-                 sig_time_profile: Optional[time_profiles.GenericProfile] = None, # Python 3.9 bug... pylint: disable=unsubscriptable-object
-                 injector: Optional[injectors.TimeDependentPsInjector] = None, # Python 3.9 bug... pylint: disable=unsubscriptable-object
-    ) -> None:
+                 source: Optional[Dict[str, float]] = None,
+                 injector: Optional[injectors.PsInjector] = None) -> None:
         """Function info...
-
         More function info...
-
         Args:
             source:
-            bg_time_profile:
-            sig_time_profile:
             injector:
         """
-        super().__init__()
-        if injector is not None:
-            self.injector = injector
-        else:
-            self.injector = injectors.TimeDependentPsInjector(source,
-                                                              bg_time_profile,
-                                                              sig_time_profile)
+        if injector is not None: self.injector = injector
+        else: self.injector = injectors.TimeDependentPsInjector(source)
```

Comment:
I think we should have the time profiles as attributes in the injector. it doesn't slow anything down or the up too much memory, and it's cleaner.

Reply:
In the injector, it does store the time_profiles. However, it will update the time_profile if you pass a new time_profile in any of the functions require a time_profile. If nothing is passed, it will just use the time profile stored in injector

Reply:
I think it's best to just let people change the time profile using injector_object.signal_time_profile = ... than to take it as arguments for each function

Reply:
I actually disagree on this one. I think if sometimes users just want to call produce_trial and see the injected events and it shouldn't be necessary for them to call self.injector.set_profile. I think we should leave it as it is right now and decide later. It is just a minor change and wouldn't affect the final result

Reply:
I'm fine with putting the change off til later. If we do that, please make sure to only use the optional time_profile arguments in tests and examples and not in any of the analysis classes.

Reply:
No problem.